src/Enum.php

```diff
@@ -2,425 +2,141 @@
-    public function getIndex(): int
-    {
-        return $this->index;
-    }
-
-    public static function getIndices(): array
-    {
-        return array_column(static::resolve(), 'index');
-    }
+        if ($name === 'label') {
+            return $this->label;
+        }
 
-    public function getValue(): string
-    {
-        return $this->value;
+        if ($name === 'value') {
+            return $this->value;
+        }
     }
 
-    public static function getValues(): array
+    public static function __callStatic(string $name, array $arguments)
     {
-        return array_column(static::resolve(), 'value');
+        return new static($name);
     }
 
-    public function getName(): string
+    public function __call($name, $arguments)
     {
-        return $this->name;
-    }
+        if (strpos($name, 'is') === 0) {
+            $other = new static(lcfirst(str_replace('is', '', $name)));
 
-    public static function getNames(): array
-    {
-        return array_keys(static::resolve());
+            return $this->equals($other);
+        }
     }
 
-    public function isAny(array $values): bool
+    public function equals(Enum ...$others): bool
     {
-        foreach ($values as $value) {
-            if ($this->isEqual($value)) {
+        foreach ($others as $other) {
+            if (
+                get_class($this) === get_class($other)
+                && $this->value === $other->value
+            ) {
                 return true;
             }
         }
 
         return false;
     }
 
-    public function isEqual($value): bool
+    protected static function values(): array
     {
-        if (is_int($value) || is_string($value)) {
-            try {
-                $value = static::make($value);
-            } catch (InvalidValueException $error) {
-                return false;
-            } catch (InvalidIndexException $error) {
-                return false;
-            }
-        }
-
-        if ($value instanceof $this) {
-            return $value->getValue() === $this->getValue();
-        }
-
-        return false;
+        return [];
     }
 
-    public function jsonSerialize(): string
+    protected static function labels(): array
     {
-        return $this->getValue();
+        return [];
     }
 
-    /**
-     * @param int|string $value
-     *
-     * @return static
-     */
-    public static function make($value): Enumerable
+    private function findDefinition($input): ?EnumDefinition
     {
-        if (! is_int($value) && ! is_string($value)) {
-            throw new TypeError(static::class.'::make() expects string|int as argument but '.gettype($value).' given');
-        }
-
-        $name = null;
-        $index = null;
-
-        if (is_int($value)) {
-            if (! static::isValidIndex($value)) {
-                throw new InvalidIndexException($value, static::class);
+        foreach (static::resolveDefinition() as $definition) {
+            if ($definition->equals($input)) {
```

Comment:
Do we really need this definition class/object? right now it's only used as data storage after resolving until instantiating an enum.
The `EnumDefinition::equals()` method is only used once - here. I'm not sure how much we really win by introducing it instead of simply having an array in the cache array. 🤔

Reply:
It's a simple value object, that's right. But I think it does serve a purpose, it simplifies the `Enum` class just a little bit, _and_ allows the developer to actually know what kind of properties are available, instead of having to remember array keys. I dedicated a chapter in Laravel beyond CRUD on this topic, if you'd like to know more about my point of view: https://stitcher.io/blog/laravel-beyond-crud-02-working-with-data

Reply:
I know this and also your point of view. 😉 
I think that we could have a good compromise if the `EnumDefinition` implements `ArrayAccess`!?
This would allow to still use some pretty helpful `array_xyz()` functions on the cache - see https://github.com/spatie/enum/pull/56#discussion_r431786607